src/engine/mod.rs

```diff
@@ -0,0 +1,697 @@
+                        format!(
+                            "Engine could not find a resolver for the custom prop: {prop_name}.",
+                            prop_name = dsp.name
+                        ),
+                        dsp.name.clone(),
+                    ),
+                    None,
+                ));
+            }
+        }
+
+        //Validate Custom Input Validator defined in WarpgrapherConfig exists as WarpgrapherValidator
+        for pwv in props_with_validator.iter() {
+            let validator_name = pwv.validator.as_ref().ok_or_else(|| {
+                Error::new(
+                    ErrorKind::ValidatorNotFound(
+                        format!(
+                            "Failed to find custom validator for prop: {prop_name}.",
+                            prop_name = pwv.name
+                        ),
+                        pwv.name.to_string(),
+                    ),
+                    None,
+                )
+            })?;
+            if !validators.contains_key(validator_name) {
+                return Err(Error::new(
+                    ErrorKind::ValidatorNotFound(
+                        format!(
+                            "Engine could not find a validator for the custom prop: {prop_name}.",
+                            prop_name = pwv.name
+                        ),
+                        pwv.name.clone(),
+                    ),
+                    None,
+                ));
+            }
+        }
+
+        // validation passed
+        Ok(())
+    }
+}
+
+/// A Warpgrapher GraphQL engine.
+///
+/// The [`Engine`] struct Juniper GraphQL service
+/// on top of it, with an auto-generated set of resolvers that cover basic CRUD
+/// operations, and potentially custom resolvers, on a set of data types and
+/// the relationships between them.  The engine includes handling of back-end
+/// communications with the chosen databse.
+///
+/// [`Engine`]: ./struct.Engine.html
+///
+/// # Examples
+///
+/// ```rust
+/// use warpgrapher::{Engine, Neo4jEndpoint};
+/// use warpgrapher::engine::config::WarpgrapherConfig;
+///
+/// let config = WarpgrapherConfig::default();
+/// let db = Neo4jEndpoint::from_env("DB_URL").unwrap();
+///
+/// let mut engine = Engine::<(), ()>::new(config, db)
+///     .build().unwrap();
+///
+/// ```
+#[derive(Clone)]
+pub struct Engine<GlobalCtx = (), ReqCtx = ()>
+where
+    GlobalCtx: 'static + Clone + Sync + Send + Debug,
+    ReqCtx: 'static + Clone + Sync + Send + Debug + WarpgrapherRequestContext,
+{
+    pub config: WarpgrapherConfig,
+    pub database: String,
+    pub pool: Option<Pool<CypherConnectionManager>>,
+    pub global_ctx: Option<GlobalCtx>,
+    pub resolvers: WarpgrapherResolvers<GlobalCtx, ReqCtx>,
+    pub validators: WarpgrapherValidators,
+    pub extensions: WarpgrapherExtensions<GlobalCtx, ReqCtx>,
+    pub version: Option<String>,
+    root_node: Option<RootRef<GlobalCtx, ReqCtx>>,
```

Comment:
If I understood yesterday's conversation correctly, part of the motivation to employ `EngineBuilder` was to avoid having these fields as options. 

Instead of passing an `Engine` to `EngineBuilder::new`, perhaps we pass the required args (ex: `WarpgrapherConfig`) and a new `Engine` is only created by `EngineBuilder::build` ?

Reply:
No longer Options